### ISSUE TITLE
Bug: batch processing async promise handling

### ISSUE THREAD

**Issue by dreamorosi** [aws-powertools/powertools-lambda-typescript#1688]
### Expected Behaviour

When using the Batch Processing utility with an async handler, customers expect to have **all the records** in the batch processed and in case of failures, the utility should report those failures.

### Current Behaviour

The current implementation of the processing function uses `Promise.all()` which takes an array of asynchronous functions (aka promises) and starts them all at once. From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#description):

> The `Promise.all()` method is one of the [promise concurrency](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#promise_concurrency) methods. It can be useful for aggregating the results of multiple promises. It is typically used when there are multiple related asynchronous tasks that the overall code relies on to work successfully — all of whom we want to fulfill before the code execution continues.

Interestingly however, the documentation points out that:

> `Promise.all()` will reject immediately upon any of the input promises rejecting. In comparison, the promise returned by [`Promise.allSettled()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) will wait for all input promises to complete, regardless of whether or not one rejects. Use allSettled() if you need the final result of every promise in the input iterable.

What this last paragraph means is that `Promise.all()` will reject, and potentially stop processing any unresolved promise, as soon as one of them rejects.

In the context of this specific utility I believe we should have use the `Promise.allSettled()` instead, which resolves when all the promises in the input array have resolved or rejected.

This is the behavior that we want because we want to be able to gather all the results (failed or successful) so that we can report partial failures.

### Code snippet

The return type of the `Promise.allSettled()` utility is slightly different than the `Promise.all()` one.

For example, let's take an input of async functions like this:

```ts
const batchOne = [
  { "id": 99, "action": "success" },
  { "id": 98, "action": "success" } 
];

const batchTwo = [
  { "id": 1, "action": "fail" },
  { "id": 2, "action": "success" },
  { "id": 3, "action": "success" } 
];
```

with a record handler like this:
```ts
const recordHandler = async (record: { id: number, action: 'success' | 'fail' }): Promise<number> => {
  if (record.action === 'fail') {
    throw new Error(`record with id ${record.id} failed`);
  }

  return record.id;
}
```

Let's now observe the behavior of the two batches with `Promise.all()`:

```ts
const results = await Promise.all(batchOne.map((record) => recordHandler(record)));
console.log(results);
/**
[
  98,
  99,
]
*/
```
For the other, it would go like this:
```ts
const results = await Promise.all(batchTwo.map((record) => recordHandler(record)));
Uncaught Error: record with id 1 failed
```

Now, using the same two batches and the same recordHandler, let's do the same but using `Promise.allSettled()`:

```ts
await Promise.allSettled(batchOne.map((record) => recordHandler(record)));
/*
[
  { status: 'fulfilled', value: 99 },
  { status: 'fulfilled', value: 98 }
]
*/
```
and for the one that has a rejection promise, it'd go like this:
```ts
await Promise.allSettled(batchTwo.map((record) => recordHandler(record)));
/*
[
  {
    status: 'rejected',
    reason: Error: record with id 1 failed
        at recordHandler
        at Array.map (<anonymous>)
  },
  { status: 'fulfilled', value: 2 },
  { status: 'fulfilled', value: 3 }
]
*/
```

Given the above, the behavior that we really want is the second and not the first, otherwise we'll never be able to process a full batch in case one of the records in the batch throws an error.

Like I mentioned, the return signature is slightly different in that `Promise.allSettled()` returns an object with the `status` which can be either `rejected` or `fulfilled` and either `reason` or `value` for failed and fulfilled promises respectively.

We should swap the implementation to this one and use the values accordingly.

### Steps to Reproduce

See steps above. You can run the codes above in the node REPL to confirm.

### Powertools for AWS Lambda (TypeScript) version

latest

### AWS Lambda function runtime

18.x

### Packaging format used

npm

**Comment by dreamorosi** [comment]
I am leaving this to you because I know you are working already on that part of the code. I hope the details above are helpful.

**Comment by am29d** [comment]
Hey Andrea,

this is a great observation, and you are correct, we need to wait for all the promises to settle and report correctly. I will change the implementation before the release.
